src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -4103,6 +4103,44 @@ void MacroAssembler::fill_words(Register base, Register cnt, Register value)
   bind(fini);
 }
 
+// Zero blocks of memory by using CBO.ZERO.
+//
+// Aligns the base address first sufficiently for CBO.ZERO, then uses
+// CBO.ZERO repeatedly for every full block.  cnt is the size to be
+// zeroed in HeapWords.  Returns the count of words left to be zeroed
+// in cnt.
+//
+// NOTE: This is intended to be used in the zero_blocks() stub.  If
+// you want to use it elsewhere, note that cnt must be >= CacheLineSize.
+void MacroAssembler::zero_dcache_blocks(Register base, Register cnt) {
+  Label initial_table_end, loop_cbo_zero;
+
+  // Align base with cache line size.
+  neg(t0, base);
+  andi(t0, t0, CacheLineSize - 1);
+
+  // t0: the number of bytes to be filled to align the base with cache line size.
+  add(base, base, t0);
+  srai(t1, t0, 3);
+  sub(cnt, cnt, t1);
+  add(t2, zr, zr);
```

Comment:
The usage of temporary registers needs to be made known to C2. You'd better pass arguments in and add effect in the ad file.

Reply:
Given it's only made to be called from `StubRoutine::zero_blocks` stub routine and `t0-t1` are temporary registers and `t2` (aka `x7`) is caller-saved, I don't understand why it needs to be made aware for C2?

Reply:
I've updated it to use `x29`, `x30`, and `x7`, and to explicit `TEMP` them in `riscv.ad`.